### ISSUE TITLE
`rgb_to_hls` returns NaN for achromatic pixels in float16, and `RandomSnow` spreads it to every gray pixel

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4571]
## Summary

`kornia.color.rgb_to_hls` returns NaN hue for every achromatic pixel (`R == G == B`: black, gray and white alike) in `float16`, and `hls_to_rgb` turns that NaN into three NaN channels. The function divides by `max - min + eps` (`kornia/color/hls.py:82`) and by `l + eps` or `2 - l + eps` (`:85`) with `eps = 1e-8`. The smallest `float16` subnormal is about `6e-8`, so `eps` rounds to `0.0`. An achromatic pixel then divides by zero, and `(g - b)` becomes `inf - inf`. `bfloat16`, `float32` and `float64` are finite for the same input.

This is the HLS counterpart of #4560 (`rgb_to_hsv`), with a larger footprint. #4560 hits only black pixels, while this hits every gray level, including in-range ones. `RandomSnow`, the one augmentation on the HLS path, returns NaN for every gray pixel of a `float16` image. A gray border, a white background or a masked region is enough. `TestRgbToHls::test_jit` and `::test_module` are already in `testing/half_precision_xfails/cpu_float16.txt`, with no issue of their own.

## Reproduction

```python
import torch, kornia.augmentation as K
from kornia.color import hls_to_rgb, rgb_to_hls

for dev in ["cpu", "mps"]:
    for dt in [torch.float16, torch.bfloat16, torch.float32]:
        if dev == "mps" and dt is torch.bfloat16:
            continue
        gray = torch.full((1, 3, 1, 1), 0.5, device=dev, dtype=dt)
        img = torch.rand(1, 3, 6, 8, device=dev, dtype=dt)
        img[..., :2, :3] = 0.25  # a 2x3 gray patch
        torch.manual_seed(0)
        snow = K.RandomSnow(p=1.0)(img)
        print(dev, dt, "rgb_to_hls(gray):", rgb_to_hls(gray).flatten().tolist(),
              "| round trip NaN:", bool(hls_to_rgb(rgb_to_hls(gray)).isnan().any()),
              "| RandomSnow NaN pixels:", int(snow.isnan().any(1).sum()), "of 48")
```

Executed on kornia main at `90650596` (torch 2.14.0, macOS arm64):

```
cpu torch.float16 rgb_to_hls(gray): [nan, 0.5, 0.0] | round trip NaN: True | RandomSnow NaN pixels: 6 of 48
cpu torch.bfloat16 rgb_to_hls(gray): [0.0, 0.5, 0.0] | round trip NaN: False | RandomSnow NaN pixels: 0 of 48
cpu torch.float32 rgb_to_hls(gray): [0.0, 0.5, 0.0] | round trip NaN: False | RandomSnow NaN pixels: 0 of 48
mps torch.float16 rgb_to_hls(gray): [nan, 0.5, 0.0] | round trip NaN: True | RandomSnow NaN pixels: 6 of 48
mps torch.float32 rgb_to_hls(gray): [0.0, 0.5, 0.0] | round trip NaN: False | RandomSnow NaN pixels: 0 of 48
```

Black and white gray levels also give NaN saturation (`rgb_to_hls` of `0.0` and `1.0` is `[nan, l, nan]` in `float16`). A chromatic pixel round-trips fine: `(0.9, 0.2, 0.4)` comes back as `(0.8999, 0.1997, 0.3994)`.

## Expected

Finite HLS for achromatic pixels in every floating dtype. Apply the #4162 rule, a floor at the dtype's smallest subnormal rather than `finfo.tiny`, or substitute a safe denominator with `torch.where` for the `max == min` and `l in {0, 1}` cases. The hue and saturation of an achromatic pixel are conventionally `0`.

## Scope

- `rgb_to_hls`, and through it `hls_to_rgb` round trips and `RandomSnow`. `rgb_to_hsv` is #4560.
- Found while reviewing #4566, which adds a `float16` warning to `RandomSnow` that links here; a fix removes it.

## Disposition

**n/a — single symbol**

Posted on behalf of @ducha-aiki by Claude (Opus 5).
